`gradle publishToMavenLocal` fails because of dependency problem with `:save-cloud-common:signJvmPublication`

- _Gradle_ version: **8.1.1**.
 - [`KT-46466`](https://youtrack.jetbrains.com/issue/KT-46466)
 - gradle/gradle#17043

The problem apparently affects only **multiplatform** _Maven_ publications:

 * ~`save-cloud-api`~
 * `save-cloud-common`
 * `save-cloud-common-js`
 * `save-cloud-common-jvm`
 * `save-cloud-common-linuxx64`
 * `save-cloud-common-macosx64`
 * ~`test-analysis-core`~

&mdash; and it's `gradle/plugins/src/main/kotlin/com/saveourtool/save/buildutils/publishing-configuration.gradle.kts` that needs to be refactored.

The problem is that, when `publish` or `publishToMavenLocal` is invoked in order to produce **4** or **5** publications at once,  the same GPG signature (`save-cloud-common-0.4.0-SNAPSHOT-javadoc.jar.asc`) may be produced y **multiple** tasks:

`:save-cloud-common:signJvmPublication`
`:save-cloud-common:signKotlinMultiplatformPublication`
`:save-cloud-common:signLinuxX64Publication`
`:save-cloud-common:signMacosX64Publication` 

&mdash; making it impossible for Gradle to figure out the proper dependency edge between the above "sign" tasks and any of the platform-specific "publish" tasks:

* `:save-cloud-common:publishJsPublicationToMavenLocal`
* `:save-cloud-common:publishJvmPublicationToMavenLocal`
* `:save-cloud-common:publishKotlinMultiplatformPublicationToMavenLocal`
* `:save-cloud-common:publishLinuxX64PublicationToMavenLocal`
* `:save-cloud-common:publishMacosX64PublicationToMavenLocal`

### Reproducible via

Any 2+ Kotlin/MP publications in a row:

 - `gradlew :save-cloud-common:clean :save-cloud-common:publishToMavenLocal`
 - `gradlew :save-cloud-common:publishJsPublicationToMavenLocal :save-cloud-common:publishJvmPublicationToMavenLocal`

### Not reproducible via
 - `gradlew :save-cloud-common:clean :save-cloud-common:publishJsPublicationToMavenLocal`
 - `gradlew :save-cloud-common:clean :save-cloud-common:publishJvmPublicationToMavenLocal`
 - `gradlew :save-cloud-common:clean :save-cloud-common:publishKotlinMultiplatformPublicationToMavenLocal`

### Workaround

The issue can be worked around by first publishing the JVM-only subprojects:

```bash
./gradlew :save-api:publishToMavenLocal :test-analysis-core:publishToMavenLocal
```

&mdash; followed by publishing each platform-specific (JS, JVM, Native) publication of `:save-cloud-common`, one at a time:

```bash
./gradlew :save-cloud-common:publishJsPublicationToMavenLocal
./gradlew :save-cloud-common:publishJvmPublicationToMavenLocal
./gradlew :save-cloud-common:publishKotlinMultiplatformPublicationToMavenLocal
./gradlew :save-cloud-common:publishLinuxX64PublicationToMavenLocal
# Mac OS X only
./gradlew :save-cloud-common:publishMacosX64PublicationToMavenLocal
```

### Gradle log

```console
FAILURE: Build failed with an exception.

* What went wrong:
A problem was found with the configuration of task ':save-cloud-common:signJvmPublication' (type 'Sign').
  - Gradle detected a problem with the following location: 'D:\Users\...\save-cloud\save-cloud-common\build\libs\save-cloud-common-0.4.0-SNAPSHOT-javadoc.jar.asc'.

    Reason: Task ':save-cloud-common:publishJsPublicationToMavenLocal' uses this output of task ':save-cloud-common:signJvmPublication' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

    Possible solutions:
      1. Declare task ':save-cloud-common:signJvmPublication' as an input of ':save-cloud-common:publishJsPublicationToMavenLocal'.
      2. Declare an explicit dependency on ':save-cloud-common:signJvmPublication' from ':save-cloud-common:publishJsPublicationToMavenLocal' using Task#dependsOn.
      3. Declare an explicit dependency on ':save-cloud-common:signJvmPublication' from ':save-cloud-common:publishJsPublicationToMavenLocal' using Task#mustRunAfter.

    Please refer to https://docs.gradle.org/8.1.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/8.1.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 1m 36s
285 actionable tasks: 15 executed, 270 up-to-date
```

--- a/gradle/plugins/src/main/kotlin/com/saveourtool/save/buildutils/PublishingConfiguration.kt
+++ b/gradle/plugins/src/main/kotlin/com/saveourtool/save/buildutils/PublishingConfiguration.kt
@@ -14,6 +14,7 @@ import org.gradle.api.Named
 import org.gradle.api.Project
 import org.gradle.api.publish.PublishingExtension
 import org.gradle.api.publish.maven.MavenPublication
+import org.gradle.api.publish.maven.tasks.AbstractPublishToMaven
 import org.gradle.api.tasks.bundling.Jar
 import org.gradle.internal.logging.text.StyledTextOutput
 import org.gradle.internal.logging.text.StyledTextOutput.Style.Failure
@@ -24,6 +25,7 @@ import org.gradle.kotlin.dsl.create
 import org.gradle.kotlin.dsl.getByType
 import org.gradle.kotlin.dsl.support.serviceOf
 import org.gradle.kotlin.dsl.withType
+import org.gradle.plugins.signing.Sign
 import org.gradle.plugins.signing.SigningExtension
 
 /**
@@ -149,6 +151,16 @@ private fun Project.configureSigningCommon(useKeys: SigningExtension.() -> Unit 
         }
         styledOut(logCategory = "signing").style(style).println(message)
         sign(*publications.toTypedArray())
+    }
+
+    /*-
+     * A Kotlin/MPP work-around for https://youtrack.jetbrains.com/issue/KT-46466.
+     *
+     * See also https://github.com/gradle/gradle/issues/17043.
+     */
+    val signingTasks = tasks.withType<Sign>()
+    tasks.withType<AbstractPublishToMaven>().configureEach {
+        dependsOn(signingTasks)
     }
 }
 

--- a/gradle/plugins/src/main/kotlin/com/saveourtool/save/buildutils/PublishingConfiguration.kt
+++ b/gradle/plugins/src/main/kotlin/com/saveourtool/save/buildutils/PublishingConfiguration.kt
@@ -15,6 +15,7 @@ import org.gradle.api.Project
 import org.gradle.api.publish.PublishingExtension
 import org.gradle.api.publish.maven.MavenPublication
 import org.gradle.api.publish.maven.tasks.AbstractPublishToMaven
+import org.gradle.api.tasks.TaskCollection
 import org.gradle.api.tasks.bundling.Jar
 import org.gradle.internal.logging.text.StyledTextOutput
 import org.gradle.internal.logging.text.StyledTextOutput.Style.Failure
@@ -158,7 +159,7 @@ private fun Project.configureSigningCommon(useKeys: SigningExtension.() -> Unit 
      *
      * See also https://github.com/gradle/gradle/issues/17043.
      */
-    val signingTasks = tasks.withType<Sign>()
+    val signingTasks: TaskCollection<Sign> = tasks.withType()
     tasks.withType<AbstractPublishToMaven>().configureEach {
         dependsOn(signingTasks)
     }

--- a/save-agent/build.gradle.kts
+++ b/save-agent/build.gradle.kts
@@ -67,6 +67,7 @@ kotlin {
             }
         }
 
+        @Suppress("UNUSED_VARIABLE")
         val jvmMain by getting {
             dependencies {
                 implementation(libs.ktor.client.apache)
@@ -74,6 +75,7 @@ kotlin {
             }
         }
 
+        @Suppress("UNUSED_VARIABLE")
         val jvmTest by getting {
             tasks.withType<Test> {
                 useJUnitPlatform()
@@ -84,12 +86,14 @@ kotlin {
             }
         }
 
+        @Suppress("UNUSED_VARIABLE")
         val linuxX64Main by getting {
             dependencies {
                 implementation(libs.ktor.client.curl)
                 implementation(libs.kotlinx.coroutines.core.linuxx64)
             }
         }
+        @Suppress("UNUSED_VARIABLE")
         val linuxX64Test by getting
     }
 

--- a/save-cloud-common/build.gradle.kts
+++ b/save-cloud-common/build.gradle.kts
@@ -64,6 +64,7 @@ kotlin {
                 }
             }
         }
+        @Suppress("UNUSED_VARIABLE")
         val jvmMain by getting {
             dependencies {
                 implementation(project.dependencies.platform(libs.spring.boot.dependencies))
@@ -89,6 +90,7 @@ kotlin {
                 api(libs.kotlinx.coroutines.reactor)
             }
         }
+        @Suppress("UNUSED_VARIABLE")
         val jvmTest by getting {
             tasks.withType<Test> {
                 useJUnitPlatform()
@@ -104,6 +106,7 @@ kotlin {
         val linuxX64Main by getting
         val macosX64Main by getting
 
+        @Suppress("UNUSED_VARIABLE")
         val nativeMain by creating {
             dependsOn(commonMain)
             linuxX64Main.dependsOn(this)

--- a/save-demo-agent/build.gradle.kts
+++ b/save-demo-agent/build.gradle.kts
@@ -29,6 +29,7 @@ kotlin {
         val macosX64Main by getting
         val linuxX64Main by getting
 
+        @Suppress("UNUSED_VARIABLE")
         val nativeMain by creating {
             macosX64Main.dependsOn(this)
             linuxX64Main.dependsOn(this)
@@ -53,6 +54,7 @@ kotlin {
         val macosX64Test by getting
         val linuxX64Test by getting
 
+        @Suppress("UNUSED_VARIABLE")
         val nativeTest by creating {
             macosX64Test.dependsOn(this)
             linuxX64Test.dependsOn(this)

